mass_mailing_sending_queue/models/mail_mass_mailing_sending.py

```diff
@@ -0,0 +1,179 @@
+        ])
+
+    @api.multi
+    def get_recipient_batch(self, res_ids):
+        self.ensure_one()
+        batch_size = self.batch_size_get()
+        already_enqueued = self.env['mail.mail.statistics'].search([
+            ('mass_mailing_sending_id', '=', self.id),
+        ])
+        set_ids = set(res_ids)
+        new_ids = list(
+            set_ids - set(already_enqueued.mapped('res_id')))[:batch_size]
+        if set(new_ids) != set_ids:
+            return new_ids
+        return res_ids
+
+    @api.multi
+    def pending_recipients(self):
+        self.ensure_one()
+        m_mailing = self.env['mail.mass_mailing'].with_context(
+            mass_mailing_sending_id=self.id)
+        return m_mailing.get_recipients(self.mass_mailing_id)
+
+    @api.multi
+    def send_mail(self):
+        for sending in self:
+            try:
+                sending.with_context(mass_mailing_sending_id=sending.id).\
+                    mass_mailing_id.send_mail()
+            except Exception as e:
+                sending._send_error(e)
+        return True
+
+    def _send_error(self, exception):
+        self.error = tools.ustr(exception)
+        self.mass_mailing_id.state = 'done'
+        self.state = 'error'
+        self.date_end = fields.Datetime.now()
+
+    def _process_enqueued(self):
+        # Create mail_mail objects not created
+        if self.pending_recipients():
+            self.send_mail()
+        # If there is no more recipient left, mark as sending
+        if not self.pending_recipients():
+            self.state = 'sending'
+            self._process_sending()
+        else:
+            self.mass_mailing_id.state = 'sending'
+
+    def _process_sending(self):
+        # Check if there is any mail_mail object not sent
+        if not self.pending_emails():
+            self.mass_mailing_id.state = 'done'
+            self.state = 'sent'
+            self.date_end = fields.Datetime.now()
+        else:
+            self.mass_mailing_id.state = 'sending'
+
+    def _process(self):
+        self.ensure_one()
+        method = getattr(self, '_process_%s' % self.state, None)
+        if method and hasattr(method, '__call__'):
+            return method()
+        return False  # pragma: no cover
+
+    @api.model
+    def sendings_running(self):
+        return self.search([
+            ('state', 'in', ('enqueued', 'sending')),
+        ])
+
+    @api.model
+    def cron(self):
+        # Process all mail.mass_mailing.sending in enqueue or sending state
+        sendings = self.sendings_running()
+        for sending in sendings:
+            _logger.info("Sending [%d] mass mailing [%d] '%s' (%s)",
+                         sending.id, sending.mass_mailing_id.id,
+                         sending.mass_mailing_id.name, sending.state)
+            sending._process()
+        return True
+
+    @api.multi
+    def _compute_pending(self):
+        for sending in self:
+            sending.pending = 0
```

Comment:
But I have to set to pending_count = 0 the sending objects not in 'enqueued' state